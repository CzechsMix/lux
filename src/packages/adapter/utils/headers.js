// @flow
import entries from '../../../utils/entries';

type ChangeType = 'SET' | 'DELETE';
type ChangeData = [string, void | string];
type HandlerChange = (type: ChangeType, data: ChangeData) => void;

class Headers extends Map<string, string> {
  constructor(value: { [key: string]: string }) {
    super(entries(value));
  }

  get(key: string): void | string {
    return super.get(String(key).toLowerCase());
  }

  has(key: string): boolean {
    return super.has(String(key).toLowerCase());
  }

  set(key: string, value: string): this {
    super.set(String(key).toLowerCase(), value);
    return this;
  }

  delete(key: string): boolean {
    return super.delete(String(key).toLowerCase());
  }
}

export class RequestHeaders extends Headers {
  // eslint-disable-next-line no-unused-vars
  set(key: string, value: string): this {
    return this;
  }

  // eslint-disable-next-line no-unused-vars
  delete(key: string): boolean {
    return false;
  }
}

export class ResponseHeaders extends Headers {
  handleChange: HandlerChange;

  constructor(handleChange: HandlerChange) {
    super({});
    this.handleChange = handleChange;
  }

  set(key: string, value: string): this {
    this.handleChange('SET', [key, value]);
    return super.set(key, value);
  }

  delete(key: string): boolean {
    this.handleChange('DELETE', [key]);
    return super.delete(key);
  }
}

Object.defineProperty(ResponseHeaders.prototype, 'handleChange', {
  writable: true,
  enumerable: false,
  configurable: false,
});
