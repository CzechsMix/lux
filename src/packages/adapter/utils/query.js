// @flow
import entries from '../../../utils/entries';
import isObject from '../../../utils/is-object';

const INT = /^\d+$/;
const CSV = /^(?:[\w\d-]+)(?:,[\w\d-]+)*$/;
const NULL = /^null$/;
const BOOL = /^(?:true|false)$/;
const DATE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}(Z|\+\d{4})$/;
const TRUE = /^true$/;
const DELIMITER = /[_-\s]+/;

type Params = {
  [key: string]: any;
};

const camelize = (source: string) => (
  source
    .split(DELIMITER)
    .reduce((result, part, idx) => {
      if (part[0]) {
        const [first] = part;

        return (
          result
          + (idx === 0 ? first.toLowerCase() : first.toUpperCase())
          + part.slice(1).toLowerCase()
        );
      }

      return result;
    }, '')
);

const fromString = (source: string): any => {
  if (INT.test(source)) {
    return Number.parseInt(source, 10);
  } else if (BOOL.test(source)) {
    return TRUE.test(source);
  } else if (NULL.test(source)) {
    return null;
  } else if (CSV.test(source)) {
    return source.split(',');
  } else if (DATE.test(source)) {
    return new Date(source);
  }

  return source;
};

export const fromObject = (source: Object): Params => (
  entries(source).reduce((obj, [k, v]) => {
    const result = obj;
    const key = camelize(k);
    let value = v;

    if (typeof value === 'string') {
      value = fromString(value);
    } else if (isObject(value)) {
      value = fromObject(value);
    }

    if (key === 'include' && Array.isArray(value)) {
      value = value.map(camelize);
    } else if (key === 'fields' && isObject(value)) {
      value = entries(value).reduce((fields, [resource, names]) => (
        Object.assign(fields, {
          [resource]: Array.isArray(names) ? names.map(camelize) : names,
        })
      ), {});
    }

    result[key] = value;

    return result;
  }, {})
);
