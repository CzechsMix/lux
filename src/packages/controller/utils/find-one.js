// @flow
import merge from '../../../utils/merge';
// eslint-disable-next-line no-unused-vars
import type { Model, Query } from '../../database';
import type { Request } from '../../request';

import paramsToQuery from './params-to-query';

/**
 * @private
 */
export default function findOne<T: Model>(
  model: Class<T>,
  req: Request
): Query<T> {
  const params = merge(req.defaultParams, req.params);
  const { id, select, include } = paramsToQuery(model, params);

  return model.find(id)
    .select(...select)
    .include(include);
}
