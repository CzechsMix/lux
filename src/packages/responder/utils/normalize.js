// @flow
import { STATUS_CODES } from '../../../constants';
import stringify from '../../../utils/stringify';

import dataFor from './data-for';

type ResponseData = {
  data: string;
  statusCode: number;
};

/**
 * @private
 */
export default function normalize(content: any): ResponseData {
  let data;
  let statusCode;

  switch (typeof content) {
    case 'boolean':
      if (content) {
        statusCode = 204;
      } else {
        statusCode = 401;
        data = dataFor(statusCode);
      }
      break;

    case 'number':
      if (STATUS_CODES.has(content)) {
        statusCode = content;
      } else {
        statusCode = 404;
      }
      data = dataFor(statusCode);
      break;

    case 'object':
      if (content instanceof Error) {
        statusCode = Number.parseInt(content.statusCode, 10) || 500;
        data = dataFor(statusCode, content);
      } else if (!content) {
        statusCode = 404;
        data = dataFor(statusCode);
      } else {
        statusCode = 200;
        data = content;
      }
      break;

    case 'undefined':
      statusCode = 404;
      data = dataFor(statusCode);
      break;

    default:
      statusCode = 200;
      data = content;
  }

  data = stringify(data);

  return {
    data,
    statusCode,
  };
}
