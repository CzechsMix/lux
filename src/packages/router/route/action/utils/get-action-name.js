// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getActionName({
  route: {
    action
  }
}: Request): string {
  return action;
}
