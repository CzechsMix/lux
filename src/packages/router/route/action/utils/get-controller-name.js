// @flow
import type { Request } from '../../../../request';

/**
 * @private
 */
export default function getControllerName({
  route: {
    controller: {
      constructor: {
        name
      }
    }
  }
}: Request): string {
  return name;
}
