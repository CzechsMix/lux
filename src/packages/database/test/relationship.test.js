// @flow
import { get, set } from '../relationship';
import { getTestApp } from '../../../../test/utils/get-test-app';

describe('module "database/relationship"', () => {
  const id = 999999;
  const ids = [0, 1, 2].map(num => id + num);
  let store;
  let subject;

  beforeAll(async () => {
    ({ store } = await getTestApp());

    await store
      .connection('posts')
      .insert({
        id,
        title: '#get() test',
        user_id: id,
      });
  });

  afterAll(async () => {
    await store
      .connection('posts')
      .del()
      .where('id', id);
  });

  beforeEach(async () => {
    subject = await store
      .modelFor('post')
      .find(id);
  });

  describe('#get()', () => {
    beforeAll(() => (
      store.connection.transaction(trx => {
        const insert = (table, data) => (
          store
            .connection(table)
            .insert(data)
            .transacting(trx)
        );

        const insertBatch = (table, data) => (
          store.connection
            .batchInsert(table, data)
            .transacting(trx)
        );

        return Promise.all([
          insert('users', {
            id,
            name: 'Test User',
            email: 'hello@postlight.com',
            password: 'password',
          }),
          insert('images', {
            id,
            url: 'https://postlight.com',
            post_id: id,
          }),
          insertBatch('tags', ids.map(val => ({
            id: val,
            name: `#get() test ${1 + (val - id)}`,
          }))),
          insertBatch('comments', ids.map(val => ({
            id: val,
            post_id: id,
            message: `#get() test ${1 + (val - id)}`,
          }))),
          insertBatch('categorizations', ids.map(val => ({
            id: val,
            tag_id: val,
            post_id: id,
          }))),
        ]);
      })
    ));

    afterAll(() => (
      store.connection.transaction(trx => {
        const del = table => (
          store
            .connection(table)
            .transacting(trx)
            .del()
        );

        return Promise.all([
          del('users').where('id', id),
          del('images').where('id', id),
          del('tags').whereIn('id', ids),
          del('comments').whereIn('id', ids),
          del('categorizations').whereIn('id', ids),
        ]);
      })
    ));

    describe('has-one relationships', () => {
      it('resolves with the correct value when present', async () => {
        const result = await get(subject, 'image');

        expect(result).toBeTruthy();

        if (result && typeof result === 'object' && !Array.isArray(result)) {
          expect(result.toJSON()).toMatchSnapshot();
        }
      });
    });

    describe('belongs-to relationships', () => {
      it('resolves with the correct value when present', async () => {
        const result = await get(subject, 'user');

        expect(result).toBeTruthy();

        if (result && typeof result === 'object' && !Array.isArray(result)) {
          expect(result.toJSON()).toMatchSnapshot();
        }
      });
    });

    describe('one-to-many relationships', () => {
      it('resolves with the correct value when present', async () => {
        const result = await get(subject, 'comments');

        expect(Array.isArray(result)).toBe(true);

        if (Array.isArray(result)) {
          expect(result.map(comment => comment.toObject())).toMatchSnapshot();
        }
      });
    });

    describe('many-to-many relationships', () => {
      it('resolves with the correct value when present', async () => {
        const result = await get(subject, 'tags');

        expect(Array.isArray(result)).toBe(true);

        if (Array.isArray(result)) {
          expect(result.map(tag => tag.toObject())).toMatchSnapshot();
        }
      });
    });
  });

  describe('#set()', () => {
    describe('has-one relationships', () => {
      let image;

      beforeAll(async () => {
        await store
          .connection('images')
          .insert({
            id,
            url: 'https://postlight.com'
          });

        image = await store
          .modelFor('image')
          .find(id);
      });

      afterEach(() => (
        store
          .connection('images')
          .del()
          .where('id', id)
      ));

      it('can add a record to the relationship', () => {
        set(subject, 'image', image);

        expect(image.toJSON()).toMatchSnapshot();
        expect(image.dirtyRelationships.get('post')).toBe(subject);
        expect(subject.dirtyRelationships.get('image')).toBe(image);
      });
    });

    describe('belongs-to relationships', () => {
      let user;

      beforeAll(async () => {
        await store
          .connection('users')
          .insert({
            id: id + 1,
            name: 'Test User',
            email: 'hello@postlight.com',
            password: 'password',
          });

        user = await store
          .modelFor('user')
          .find(id + 1);
      });

      afterEach(() => (
        store
          .connection('users')
          .del()
          .where('id', id + 1)
      ));

      it('can add a record to the relationship', () => {
        set(subject, 'user', user);

        expect(subject.toJSON()).toMatchSnapshot();
        expect(subject.dirtyRelationships.get('user')).toBe(user);
      });
    });

    describe('one-to-many relationships', () => {
      let comments;

      beforeAll(() => (
        store.connection.batchInsert('comments', ids.map(val => ({
          id: val,
          message: `#set() test ${1 + (val - id)}`,
        })))
      ));

      afterAll(() => (
        store
          .connection('comments')
          .del()
          .whereIn('id', ids)
      ));

      beforeEach(async () => {
        comments = await store
          .modelFor('comment')
          .select('id', 'edited', 'message')
          .where({
            id: ids,
          });
      });

      it('can add records to the relationship', () => {
        set(subject, 'comments', comments);

        const dirty = subject.dirtyRelationships.get('comments');

        expect(Array.isArray(dirty)).toBe(true);
        dirty.forEach(expect(comments).toContain);

        expect(comments.map(comment => comment.toJSON())).toMatchSnapshot();
        comments.forEach(comment => {
          expect(comment.dirtyRelationships.get('post')).toBe(subject);
        });
      });

      it('can remove records from the relationship', () => {
        set(subject, 'comments', []);

        expect(subject.dirtyRelationships.get('comments')).toEqual([]);
        expect(subject.toJSON()).toMatchSnapshot();
      });
    });
  });
});
