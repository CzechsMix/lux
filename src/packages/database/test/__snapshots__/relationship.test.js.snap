exports[`module "database/relationship" #get() belongs-to relationships resolves with the correct value when present 1`] = `
Object {
  "createdAt": null,
  "email": "hello@postlight.com",
  "id": 999999,
  "name": "Test User",
  "password": "password",
  "updatedAt": null,
}
`;

exports[`module "database/relationship" #get() has-one relationships resolves with the correct value when present 1`] = `
Object {
  "createdAt": null,
  "id": 999999,
  "postId": 999999,
  "updatedAt": null,
  "url": "https://postlight.com",
}
`;

exports[`module "database/relationship" #get() many-to-many relationships resolves with the correct value when present 1`] = `
Array [
  Object {
    "createdAt": null,
    "id": 999999,
    "name": "#get() test 1",
    "updatedAt": null,
  },
  Object {
    "createdAt": null,
    "id": 1000000,
    "name": "#get() test 2",
    "updatedAt": null,
  },
  Object {
    "createdAt": null,
    "id": 1000001,
    "name": "#get() test 3",
    "updatedAt": null,
  },
]
`;

exports[`module "database/relationship" #get() one-to-many relationships resolves with the correct value when present 1`] = `
Array [
  Object {
    "createdAt": null,
    "edited": false,
    "id": 999999,
    "message": "#get() test 1",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
  Object {
    "createdAt": null,
    "edited": false,
    "id": 1000000,
    "message": "#get() test 2",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
  Object {
    "createdAt": null,
    "edited": false,
    "id": 1000001,
    "message": "#get() test 3",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
]
`;

exports[`module "database/relationship" #set() belongs-to relationships can add a record to the relationship 1`] = `
Object {
  "body": null,
  "createdAt": null,
  "id": 999999,
  "isPublic": false,
  "title": "#get() test",
  "updatedAt": null,
  "userId": 1000000,
}
`;

exports[`module "database/relationship" #set() has-one relationships can add a record to the relationship 1`] = `
Object {
  "createdAt": null,
  "id": 999999,
  "postId": 999999,
  "updatedAt": null,
  "url": "https://postlight.com",
}
`;

exports[`module "database/relationship" #set() one-to-many relationships can add records to the relationship 1`] = `
Array [
  Object {
    "createdAt": null,
    "edited": false,
    "id": 999999,
    "message": "#set() test 1",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
  Object {
    "createdAt": null,
    "edited": false,
    "id": 1000000,
    "message": "#set() test 2",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
  Object {
    "createdAt": null,
    "edited": false,
    "id": 1000001,
    "message": "#set() test 3",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
]
`;

exports[`module "database/relationship" #set() one-to-many relationships can remove records from the relationship 1`] = `
Array [
  Object {
    "createdAt": null,
    "edited": false,
    "id": 999999,
    "message": "#set() test 1",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
  Object {
    "createdAt": null,
    "edited": false,
    "id": 1000000,
    "message": "#set() test 2",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
  Object {
    "createdAt": null,
    "edited": false,
    "id": 1000001,
    "message": "#set() test 3",
    "postId": 999999,
    "updatedAt": null,
    "userId": null,
  },
]
`;
